homeassistant/util/logging.py

```diff
@@ -2,21 +2,93 @@
 
 from __future__ import annotations
 
+from collections import defaultdict
 from collections.abc import Callable, Coroutine
+import datetime
 from functools import partial, wraps
 import inspect
 import logging
 import logging.handlers
-import queue
+from queue import SimpleQueue
 import traceback
-from typing import Any, cast, overload
+from typing import Any, cast, overload, override
 
 from homeassistant.core import (
     HassJobType,
     HomeAssistant,
     callback,
     get_hassjob_callable_job_type,
 )
+from homeassistant.helpers.event import async_track_time_interval
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class HomeAssistantQueueListener(logging.handlers.QueueListener):
+    """Custom QueueListener to watch for noisy loggers."""
+
+    LOG_COUNTS_RESET_INTERVAL = datetime.timedelta(minutes=5)
+    MAX_LOGS_COUNT = 200
+
+    _reset_pending: bool
+    _log_counts: dict[str, int]
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        queue: SimpleQueue[logging.Handler],
+        *handlers: logging.Handler,
+    ) -> None:
+        """Initialize the handler."""
+        super().__init__(queue, *handlers)
+        self._hass = hass
+        self._reset_counters()
+
+        @callback
+        def _set_reset_pending_flag(_: datetime.datetime) -> None:
+            _LOGGER.debug("Setting reset pending")
+            # Set a boolean instead of directly resetting the counters so that the actual
+            # reset happens on the QueueListener thread without the need for a Lock.
+            self._reset_pending = True
+
+        async_track_time_interval(
+            self._hass,
+            _set_reset_pending_flag,
+            self.LOG_COUNTS_RESET_INTERVAL,
+            cancel_on_shutdown=True,
+        )
+
+    @override
+    def handle(self, record: logging.LogRecord) -> None:
+        """Handle the record."""
+        super().handle(record)
+
+        if record.levelno < logging.INFO:
+            return
+
+        if self._reset_pending:
+            self._reset_counters()
+
+        module_name = record.name
+        if module_name == __name__:
+            return
+
+        self._log_counts[module_name] += 1
+        module_count = self._log_counts[module_name]
+        if module_count < self.MAX_LOGS_COUNT:
+            return
+
+        _LOGGER.warning(
+            "Module %s is logging too frequently. %d messages since last count",
```

Comment:
Should we keep track of modules we have warned for and only warn once? `self._seen = set()` 

No need to spam more warnings

Reply:
I had that on the very beginning but removed it because if the "bad" logs keep spaming, the warning would be "lost" in the big log file.
OTOH, I agree that this would add to the spam and since we already group logs on the UI, I guess it is fine to log only once.